multicluster/controllers/multicluster/node_controller.go

```diff
@@ -73,65 +78,193 @@ func NewNodeReconciler(
+			isValidGateway = false
 		}
 	}
-	if err := r.Client.Get(ctx, gwNamespacedName, gw); err != nil {
-		if apierrors.IsNotFound(err) && isGW {
-			gw.GatewayIP = gwIP
-			gw.InternalIP = internalIP
-			if err := r.Client.Create(ctx, gw, &client.CreateOptions{}); err != nil {
+
+	if isActiveGateway {
+		if !stillGatewayNode || !isValidGateway || !isReadyNode(node) {
+			if err := r.recreateActiveGateway(ctx, gw); err != nil {
 				return ctrl.Result{}, err
 			}
 			return ctrl.Result{}, nil
 		}
-		return ctrl.Result{}, client.IgnoreNotFound(err)
+		if err := r.updateActiveGateway(ctx, gw); err != nil {
+			return ctrl.Result{}, err
+		}
+	}
+
+	if noActiveGateway && isValidGateway && stillGatewayNode && isReadyNode(node) {
+		if err := r.createGateway(gw); err != nil {
+			return ctrl.Result{}, err
+		}
 	}
+	return ctrl.Result{}, nil
+}
 
-	if !isGW {
-		err := r.Client.Delete(ctx, gw, &client.DeleteOptions{})
-		return ctrl.Result{}, client.IgnoreNotFound(err)
+// initialize initializes 'activeGateway' and 'gatewayCandidates' and removes
+// stale Gateway during controller startup.
+func (r *NodeReconciler) initialize() error {
+	ctx := context.Background()
+	nodeList := &corev1.NodeList{}
+	if err := r.Client.List(ctx, nodeList, &client.ListOptions{}); err != nil {
+		return err
 	}
 
-	gw.GatewayIP = gwIP
-	gw.InternalIP = internalIP
-	if err := r.Client.Update(ctx, gw, &client.UpdateOptions{}); err != nil {
-		return ctrl.Result{}, err
+	gwList := &mcsv1alpha1.GatewayList{}
+	if err := r.Client.List(ctx, gwList, &client.ListOptions{}); err != nil {
+		return err
 	}
-	return ctrl.Result{}, nil
+	// Gateway webhook guarantees that there is at most one Gateway in the member cluster.
+	if len(gwList.Items) > 0 {
+		existingGWName := gwList.Items[0].Name
+		node := &corev1.Node{}
+		if err := r.Client.Get(ctx, types.NamespacedName{Name: existingGWName}, node); err != nil {
+			if !apierrors.IsNotFound(err) {
+				return err
+			}
+			staleGateway := &mcsv1alpha1.Gateway{
+				ObjectMeta: metav1.ObjectMeta{
+					Namespace: r.namespace,
+					Name:      existingGWName},
+			}
+			err := r.Client.Delete(ctx, staleGateway, &client.DeleteOptions{})
+			if err != nil && !apierrors.IsNotFound(err) {
+				return err
+			}
+		} else {
+			r.activeGateway = existingGWName
+		}
+	}
+	for _, n := range nodeList.Items {
+		if _, isGW := n.Annotations[common.GatewayAnnotation]; isGW {
+			r.gatewayCandidates[n.Name] = true
+		}
+	}
+	return nil
+}
+
+func (r *NodeReconciler) updateActiveGateway(ctx context.Context, newGateway *mcsv1alpha1.Gateway) error {
+	existingGW := &mcsv1alpha1.Gateway{}
+	// TODO: cache might be stale. Need to revisit here and other reconcilers to
+	// check if we can improve this with 'Owns' or other methods.
+	if err := r.Client.Get(ctx, types.NamespacedName{Name: newGateway.Name, Namespace: r.namespace}, existingGW); err != nil {
+		return client.IgnoreNotFound(err)
+	}
+	existingGW.GatewayIP = newGateway.GatewayIP
```

Comment:
I think we should check, and if no change just return, to avoid an update for any Node change?